server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java

```diff
@@ -169,6 +171,13 @@ public <T> T getAggregator(RegistryKey<T> registryKey, ValuesSourceConfig values
             @SuppressWarnings("unchecked")
             T supplier = (T) aggregatorRegistry.get(registryKey).get(valuesSourceConfig.valueSourceType());
             if (supplier == null) {
+                if (valuesSourceConfig.fieldType() != null
+                    && valuesSourceConfig.fieldType().typeName() != null
+                    && "aggregate_metric_double".equals(valuesSourceConfig.fieldType().typeName().toLowerCase(Locale.ROOT))) {
+                    throw new UnsupportedAggregationOnDownsampledField(
+                        valuesSourceConfig.getDescription() + " is not supported for aggregation [" + registryKey.getName() + "]"
+                    );
+                }
```

Comment:
I wonder if there's a better place to put this. It feels like kind of a janky spot. @not-napoleon, do you have an opinion on this? Basically we want field types to be able to throw a specific exception if you ask or an unsupported aggregation. Well, really, we just want *this* field to throw it, but I'll bet there is a cleaner way.

Reply:
Maybe I can do this at the moment the aggregator is created.

Reply:
This needs to be done for all aggregations. So the only place I see to do it (excluding doing it in every aggregation) is some base class where the construction of the aggregator takes place. Any better idea?

Reply:
Notice, anyway, that if I remove this piece of code, it fails anyway with 

`throw new IllegalArgumentException(
                    valuesSourceConfig.getDescription() + " is not supported for aggregation [" + registryKey.getName() + "]"
);`

so at the end that is just a way to replace one exception with another.

An option I see is to extract the responsibility of checking for "no value in the registry" with another class that is injected into the ValuesSourceRegistry. A kind of "no such element in the registry" handler.

Reply:
I think we already have the handler injected - the `ValuesSourceType` itself can do this.  What you're asking for here is a type-specific behavior.  In aggregations, we don't (generally) work directly with field types, we instead work with the `ValuesSourceType` abstraction, so unless we have a Real Good Reason :TM: that should be the lowest granularity of type switching (real good reasons generally involve "we can optimize this specific index structure, but only if we know about it", for reference).

So, I'd suggest we add a method to the `ValuesSourceType` interface `getUnmappedException(String aggregationName)` or something like that (the VST already has a `typeName()` method it can use for the other string component).  We can add a default implementation that returns the current exception, and override it in `AggergateMetricValuesSourceType.AGGREGATE_METRIC` to throw your new exception.  This also has the advantage of locating the special messaging close to the associated type logic, which seems more natural than hard coding it in the registry.

Let me know if that doesn't work for some reason.

Reply:
That sound better. Thanks.

Reply:
I had to add the `ValueSourceConfig` to the parameters because otherwise composing the exact error message was not possible. The `getDescription` has some logic to build it.

Reply:
I wouldn't bother to check that the exception is non-null here. Returning a `null` supplier isn't really a thing we want to support. I think it's an error is the type doesn't return an exception.

Reply:
this needs an else-throw.  If we didn't find a supplier in the registry, we *MUST* throw from this method, or something very weird will break down the road.  Really, it should be impossible for `getUnmappedException` to return `null`, but I'm all for redundant checks.

Reply:
I see...I thought about that but my idea was "there is the default method". Anyway that is not enough. A subclass overriding the parent method might still return null and then the flow would continue without an exception thrown.

Reply:
Does it make sense to add a warning message saying something like "returned unexpected null exception, please fix this!"?

Reply:
You'd asked about double checking that this isn't null rather than just throwing blindly. I'm ok either way. It's a "you built this wrong" sort of error. I don't expect folks to hit it much, but it'd be kind to tell them. I'd use an `assert` for it or not build it at all. But you can do whatever you like here. no big deal to me.

Reply:
yeah, thinking about it more, the assert is probably the right choice.  We want to catch this in testing, not give the user an error they can't do anything with, and the assert communicates that more clearly.